src/coreclr/vm/gchelpers.cpp

```diff
@@ -339,12 +407,52 @@ void PublishObjectAndNotify(TObj* &orObject, GC_ALLOC_FLAGS flags)
     {
         ETW::TypeSystemLog::SendObjectAllocatedEvent(orObject);
     }
+
+    // TODO: not sure it is worth emitting the event here instead of in Alloc
+    //       because it is needed to pass the size + the isSampled flag
+    //       so all caller of PublishObjectAndNotify should pass the size and the isSampled flag
+    if (isSampled && EventPipeEventEnabledAllocationSampled())
+    {
+        // TODO: this code is duplicated from GCToCLREventSink::FireGCAllocationTick_V4
+        //       --> should we refactor into a common helper?
+        void* typeId = nullptr;
+        const WCHAR* name = nullptr;
+        InlineSString<MAX_CLASSNAME_LENGTH> strTypeName;
+        EX_TRY
+        {
+            TypeHandle th = GetThread()->GetTHAllocContextObj();
+
+            if (th != 0)
+            {
+                th.GetName(strTypeName);
+                name = strTypeName.GetUnicode();
+                typeId = th.GetMethodTable();
+            }
+        }
+        EX_CATCH{}
+        EX_END_CATCH(SwallowAllExceptions)
+        // end of duplication
+
+        if (typeId != nullptr)
+        {
+            unsigned int allocKind =
+                (flags & GC_ALLOC_PINNED_OBJECT_HEAP) ? 2 :
+                (flags & GC_ALLOC_LARGE_OBJECT_HEAP) ? 1 :
+                0;  // SOH
+            unsigned int heapIndex = 0;
+#ifdef BACKGROUND_GC
+            gc_heap* hp = gc_heap::heap_of((BYTE*)orObject);
+            heapIndex = hp->heap_number;
+#endif
+            FireEtwAllocationSampled(allocKind, GetClrInstanceId(),typeId, name, heapIndex, (BYTE*)orObject, size);
+        }
+    }
 #endif // FEATURE_EVENT_TRACE
 }
 
-void PublishFrozenObject(Object*& orObject)
+void PublishFrozenObject(Object*& orObject, size_t size)
 {
-    PublishObjectAndNotify(orObject, GC_ALLOC_NO_FLAGS);
+    PublishObjectAndNotify(orObject, size, GC_ALLOC_NO_FLAGS, false);  // TODO: allocations in NGCH are not sampled?
```

Comment:
It makes sense to me that we'd sample it for consistency, but we probably need to define a new value for the allocKind enumeration.

Reply:
I was not sure because there is no public API to allocate there

Reply:
At the moment all allocations on the non-gc heap are implicit. I think all string constants get allocated there.

Reply:
Yes: string constants, runtime types and others but never for objects allocated by user code. Also, the undocumented API to register "frozen" segments does not trigger object allocation callback so they will be missed and the sampling won't apply. This is why I don't think the event should be raised.